src/geometry/VBO_Handler.cxx

```diff
@@ -0,0 +1,607 @@
+#endif
+
+// common implementation header
+#include "OpenGLGState.h"
+
+bool VBO_Handler::textureEnabled = false;
+bool VBO_Handler::normalEnabled  = false;
+bool VBO_Handler::colorEnabled   = false;
+GLuint VBO_Handler::actVerts     = 0;
+GLuint VBO_Handler::actNorms     = 0;
+GLuint VBO_Handler::actTxcds     = 0;
+GLuint VBO_Handler::actColrs     = 0;
+
+bool VBO_Manager::glContextReady = false;
+VBO_Manager::VBO_Manager ()
+{
+    OpenGLGState::registerContextInitializer(freeContext, initContext, this);
+}
+
+VBO_Manager::~VBO_Manager ()
+{
+    OpenGLGState::unregisterContextInitializer(freeContext, initContext, this);
+}
+
+void VBO_Manager::registerHandler(VBO_Handler *handler)
+{
+    handlerList.push_back(handler);
+}
+
+void VBO_Manager::unregisterHandler(VBO_Handler *handler)
+{
+    std::list <VBO_Handler*> :: iterator it;
+    for (it = handlerList.begin(); it != handlerList.end(); it++)
+        if (*it == handler)
+        {
+            handlerList.erase(it);
+            break;
+        }
+}
+
+void VBO_Manager::registerClient(VBOclient *client)
+{
+    clientList.push_back(client);
+    if (glContextReady)
+        client->initVBO();
+}
+
+void VBO_Manager::unregisterClient(VBOclient *client)
+{
+    std::list <VBOclient*> :: iterator it;
+    for (it = clientList.begin(); it != clientList.end(); it++)
+        if (*it == client)
+        {
+            clientList.erase(it);
+            break;
+        }
+}
+
+void VBO_Manager::initAll()
+{
+    std::list <VBO_Handler*>::const_iterator itH;
+    for (itH = handlerList.begin(); itH != handlerList.end(); itH++)
+        (*itH)->init();
+    std::list <VBOclient*>::const_iterator itC;
+    for (itC = clientList.begin(); itC != clientList.end(); itC++)
+        (*itC)->initVBO();
+}
+
+void VBO_Manager::destroyAll()
+{
+    std::list <VBO_Handler*>::const_iterator it;
+    for (it = handlerList.begin(); it != handlerList.end(); it++)
+        (*it)->destroy();
+}
+
+void VBO_Manager::initContext(void* data)
+{
+    glContextReady = true;
+    ((VBO_Manager*)data)->initAll();
+}
+
+void VBO_Manager::freeContext(void* data)
+{
+    glContextReady = false;
+    ((VBO_Manager*)data)->destroyAll();
+}
+
+void VBO_Manager::reset ()
+{
+    destroyAll();
+    initAll();
+}
+
+VBO_Handler::VBO_Handler(
+    bool _handleTexture,
+    bool _handleNormal,
+    bool _handleColor,
+    int _vertexSize,
+    int _indexSize) :
+    handleTexture(_handleTexture), handleNormal(_handleNormal),
+    handleColor(_handleColor), vertexSize(_vertexSize), indexSize(_indexSize),
+    arrayFillPoint(0), indexFillPoint(0)
+{
+    vboManager.registerHandler(this);
+    freeVBOList.push_back({0, vertexSize});
+}
+
+VBO_Handler::~VBO_Handler()
+{
+    vboManager.unregisterHandler(this);
+}
+
+void VBO_Handler::init()
+{
+    glGenBuffers(1, &verts);
+    glBindBuffer(GL_ARRAY_BUFFER, verts);
+    glBufferData(
+        GL_ARRAY_BUFFER,
+        vertexSize * sizeof(GLfloat[3]),
+        NULL,
```

Comment:
even if glBufferData is a C procedure? probably NULL and nullptr are both 0, but doc says to pass NULL

Reply:
in C, there is no compiler keyword for a null pointer. The language **does** guarantee that a value of 0 assigned to a pointer will be a null pointer, regardless of the underlying implementation. `NULL` is a macro that was originally defined as `(char*)0` but is now (I believe) defined as `(void*)0`.

In C++11, the `nullptr` keyword was added to explicitly define a null pointer value. Since you are writing C++, you can (and should) confidently use `nullptr` any place a null pointer is needed.